### DISCUSSION TITLE
Issue with PMS9103M driver

### DISCUSSION THREAD

**Discussion by ghost** [micropython/micropython#15637] [ESP32]
I've been trying to get the asynchronous driver I previously made for the PMS7003 particle sensor to work with the PMS9103M, but for some strange reason, I usually only get zero values, or just the number of particles, but never the PM values (ATM) needed for calculating the air quality index. 

In the past, these types of problems have been related to UART issues, but I believe the UART is correctly set up in my code. The corresponding initialization works for another UART with the MH-Z19B CO2 sensor, and it functions correctly. I tried adding commands to force the sensor into active mode and other unnecessary things. Has anyone made a working driver for the PMS9103M that I could test? 

The driver I have build: https://github.com/divergentti/Micropython-for-ESP32/blob/main/esp32-oled-mhz19b-pms9103m-bme680/drivers/PMS9103M_AS.py

Whole program here, everything else but PMS readings works https://github.com/divergentti/Micropython-for-ESP32/tree/main/esp32-oled-mhz19b-pms9103m-bme680

Debug screen:

RAM free 80736 alloc 68128

1 ---------WIFI------------- 1
   WiFi Connected True, hotspot: hidden, signal strength: -54
   IP-address: 192.168.x.x
   Runtime: 607
   MQTT Connected: True, broker uptime: b'10901 seconds'
   MQTT messages sent 30 seconds ago.
   Memory free: 80288, allocated: 68576
2 -------SENSORDATA--------- 2
   Temp: 34.0C, Rh: 31.4, GasR: 726279.8
   CO2 is 631.0
   PM1:0 (0) PM2.5:0 (0)
   PM10: 0 (ATM: 0)
   0 < 0.3 & 0 <0.5
   0 < 1.0 & 0 < 2.5
   0 < 5.0 & 0 < 10.0
3 ---------FAULTS------------- 3
   Last error : None
   Read errors: 0

PMS reader data b'B'  (this is START_BYTE_1 = 0x42)
PMS reader data b'M' (this is START_BYTE_2 = 0x4d)
PMS reader data b'\x00\x1c\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x13\x00\x00\xbe'

**Comment by rahulatchuta** [comment]
Hello,
I'm also facing the same problem with the PMS9103M,  even i couldn't get the PM measurement results instead I'm getting zeros like you have received.
Here is my output.
Sending command: [66, 77, 228, 0, 1, 1, 116]
Sending command: [66, 77, 226, 0, 1, 1, 113]
Raw frame (Hex): 424d001c0000000000000000000000000000003a000000000000000020000105
Calculated checksum: 261, Received checksum: 261
Parsed Data: {'FRAME_LENGTH': 28, 'PM1_0': 0, 'PM2_5': 0, 'PM10_0': 0, 'PM1_0_ATM': 0, 'PM2_5_ATM': 0, 'PM10_0_ATM': 0, 'PCNT_0_3': 0, 'PCNT_0_5': 58, 'PCNT_1_0': 0, 'PCNT_2_5': 0, 'PCNT_5_0': 0, 'PCNT_10_0': 0, 'VERSION': 32, 'ERROR': 0, 'CHECKSUM': 261} 

I have sent two commands first one to wakeup the sensor and second to read the measurement results in active mode.

If you have found out the solution to this issue please share it with me.

**Comment by ghost** [comment]
Hello there. I have not yet solved this problem. I bought another sensor from Aliexpress and it worked  opposite. I had conversation with manufacturer and they asked me to blow cigarette smoke to the sensor. I do not smoke, but I tested with candle smoke and after that I got reading. However, PMS7003 works like a charm without smoke: https://github.com/divergentti/Micropython-for-ESP32/blob/main/Esp-Drivers/Sensors/PMS7003_AS.py  ... so perhaps PMS9103M is for some other purposes.

Following is the manufacturer's  response:

Dear Jari，


Please try to light the cigarette is there any change, please see the fan running?
Please check.

-----

    Hello.

    Is it possible that Aliexpress sells some sort of fake sensors? I bought new and this one is giving:

    Data received: bytearray(b'B')
    Data received: bytearray(b'M')
    Data received: bytearray(b'\x00\x1c\x00\x02\x00\x04\x00\x05\x00\x02\x00\x04\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc1')
    Raw data received: bytearray(b'\x00\x1c\x00\x02\x00\x04\x00\x05\x00\x02\x00\x04\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc1')
    PMS Read at 778944757, data: {'PM2_5': 4, 'PM1_0_ATM': 2, 'PM10_0_ATM': 5, 'VERSION': 0, 'ERROR': 0, 'CHECKSUM': 193, 'PM1_0': 2, 'PM10_0': 5, 'PM2_5_ATM': 4, 'PCNT_0_5': 0, 'PCNT_5_0': 0, 'FRAME_LENGTH': 28, 'PCNT_2_5': 0, 'PCNT_1_0': 0, 'PCNT_10_0': 0, 'PCNT_0_3': 0}

    So, frame is correct, but now PM-values are low, keeping low, but no PCNT at all. 

    To my understanding the UART port should work ok, if it communicates with PMS9103M.

    Thank you for your advice.

----

    Dear  Jari Hiltunen,

    For some reason, the data you mentioned is always one more byte.

    You can light a cigarette to see if the sensor data changes, to see if the fan is running.

----

            Hello and thank you for your prompt answer.

            Here is a little bit more information. Datasheet I am reading is[ https://evelta.com/content/datasheets/203-PMS9003M.pdf](https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fevelta.com%2Fcontent%2Fdatasheets%2F203-PMS9003M.pdf&data=05%7C02%7C%7C4505e502776143bf644508dcd07e772f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638614488821871919%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=qqei31Iu9g6e6Uj7gtQLpGs87mI7pEtQHZMaD%2BiXjrE%3D&reserved=0)

            Simple python test:

            import machine
            from machine import UART
            sensor = UART(2, baudrate=9600, bits=8, parity=None, stop=1, rx=Pin(16), tx=Pin(17))

            In active mode I receive:
            >>> sensor.readline()
            b'BM\x00\x1c\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x80\x00j\x00\x00\x00\x00\x00\x00\x00\x00\x13\x00\x01\xa8'

            and the sensor keeps sending these values continously. If I set sensor to passive mode with command:
            >>> sensor.write(bytearray([0x42, 0x4d, 0xe1, 0x00, 0x00, 0x01, 0x70]))

            I receive response (I guess response 5 means fail):
            7

            >>>
            and if I read with passive mode read command:
            >>> sensor.write(bytearray([0x42, 0x4d, 0xe2, 0x00, 0x00, 0x01, 0x71]))

            I receive response:
            7

            and now I can read sensor data with:

            >>> sensor.readline()
            b'BM\x00\x1c\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x96\x00Q\x00\x00\x00\x00\x00\x00\x00\x00\x13\x00\x01\xa5'

            Here we have letter Q which is 0x51h (81 decimal),

            This translates to bytearray:

            # Bytearray without two last bytes (which are H and L checksum)
            data = bytearray([0x42, 0x4d, 0x00, 0x1c, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                              0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x96, 0x00, 0x51,
                              0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x13, 0x00])

            # Sum all values
            total_sum = sum(data)

            # Calculate Low byte (L) and High byte (H)
            L = total_sum & 0xFF
            H = (total_sum >> 8) & 0xFF

            print("High byte is %s and low byte is %s" %(H,L))

            High byte is 1 and low byte is 165

            So, if we look at the CRC the sensor sent (x01\xa5) translates this in decimal 1 and 165. So, the CRC is correct.

            However, all measurement values are typically all zeroes, sometimes I see <0.3 and <0.5 particle counters small values, which seems to decrease together.

            And that does not change after 30 minutes running:

            >>> sensor.write(bytearray([0x42, 0x4d, 0xe2, 0x00, 0x00, 0x01, 0x71]))
            7
            >>> sensor.readline()
            b'BM\x00\x1c\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x87\x00Y\x00\x00\x00\x00\x00\x00\x00\x00\x13\x00\x01\x9e'

            To me this looks like the serial port works ok, the sensor responses to commands OK, but readings the sensor sends are wrong. Third byte should be error byte and that is 0.

            Do you have better command preference than in the datasheet? Or perhaps some ways to verify that the sensor is broken?

**Comment by rahulatchuta** [comment]
Hi there, Thank you so much for your quick reply. The information you have provided is really helpful.

Best regards
Rahul
